kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -110,6 +114,32 @@ public TransactionBuilder withMaxRetries(int maxRetries) {
     return this;
   }
 
+  @Override
+  public TransactionBuilder withDomainMetadata(String domainName, String config) {
+    checkArgument(
+        !domainName.toLowerCase(Locale.ROOT).startsWith("delta."),
+        "Setting a system-controlled domain is not allowed: " + domainName);
+    checkArgument(
+        !domainMetadatasRemoved.contains(domainName),
+        "Cannot add a domain that is removed in this transaction");
+    // we override any existing value
+    domainMetadatasAdded.put(
+        domainName, new DomainMetadata(domainName, config, false /* removed */));
+    return this;
+  }
+
+  @Override
+  public TransactionBuilder withDomainMetadataRemoved(String domainName) {
+    checkArgument(
+        !domainName.toLowerCase(Locale.ROOT).startsWith("delta."),
```

Comment:
probably easier if we create a DomainName class to encapsualate the case insensitive and functions like isSystemDomain etc.

Reply:
yeah I think maybe for now these can just be a static method in DomainMetadata `isSystemDomain` and `isUserDomain`